src/document.rs

```diff
@@ -282,6 +379,67 @@ impl fmt::Debug for Document {
     }
 }
 
+impl Serialize for Document {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut map =
+            serializer.serialize_map(Some(self.field_values.len()))?;
+        for (k, v) in &self.field_values {
+            let ser_v: Vec<_> =
+                v.iter().cloned().map(SerdeValue::from).collect();
+            map.serialize_entry(&k, &ser_v)?;
+        }
+        map.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for Document {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct MapVisitor;
+
+        impl<'de> Visitor<'de> for MapVisitor {
```

Comment:
Wouldn't it be easier to just deserialize `BTreeMap<String, Vec<SerdeValue>>` and then convert this to `BTreeMap<String, Vec<Value>>`?

Reply:
(I think the re-allocation could even be avoid if the conversion uses `.into_iter().map(..).collect()` as the types should layout compatible.)

Reply:
These are great suggestions, done.